internal/pkg/template/workload.go

```diff
@@ -440,7 +469,20 @@ func envControllerParameters(o WorkloadOpts) []string {
 	return parameters
 }
 
-// ARN determines the arn for a topic using the SNSTopic arn start and the name of the topic
+// ARN determines the arn for a topic using the SNSTopic name and account information
 func (t Topic) ARN() string {
 	return fmt.Sprintf(snsArnPattern, t.Partition, t.Region, t.AccountID, t.App, t.Env, t.Svc, aws.StringValue(t.Name))
 }
+
+// URI determines the uri for a queue using the queue name and account information
+func (q SQSQueue) URI(topicName *string) string {
+	var name string
+	if q.Name != nil {
+		name = aws.StringValue(q.Name)
+	} else if topicName != nil {
+		name = aws.StringValue(topicName)
+	} else {
+		name = "SQSQueue"
+	}
+	return fmt.Sprintf(sqsURIPattern, q.URL, q.AccountID, q.App, q.Env, q.Svc, name)
```

Comment:
Will this URL be constructable in the template? If yes, this may probably alleviate the parameter drilling (i.e. passing parameters deep down only because the functions at the end need it) that we have right now.

Reply:
Yes and no. We need to pass the queue URLs into a JSON environment variable, which is difficult. The other reason is security--for sns topics, we need to validate that the constructed topic name for the subscription (with app, env, service, and topic name) exactly matches one of the topics previously deployed by copilot. But otherwise policies or resources which need queue ARNs or URLs as inputs can get them by calling !Ref.